File: 算法设计与分析作业/3.19 快速排序.cpp
#include <iostream>
#include <cstdio>
#include <sstream>
using namespace std;

void swap(int &a,int &b)
{
	if(a==b) return;
	a=a^b;
	b=a^b;
	a=a^b;
}

int partition(int a[],int l,int r)//¸ô°å 
{
	int i=l-1;
	for(int j=l;j<r;j++)
		if(a[j]<a[r])
		{
			i++;
			swap(a[i],a[j]);
		}
	swap(a[i+1],a[r]);
	return i+1;
}
void quicksort(int a[],int l,int r,int &time)//¿ìÅÅ 
{
	time++;
	if(l<r)
	{
		int m=partition(a,l,r);
		quicksort(a,l,m-1,time);
		quicksort(a,m+1,r,time);
	}
}

int main()
{
	ios::sync_with_stdio(0),cin.tie(0),cout.tie(0);
	stringstream ss;
	string s;
	getline(cin,s);
	ss<<s;
	int n=0,t,a[1005],time=0;
	while(ss>>t)
		a[n++]=t;
	quicksort(a,0,n-1,time);
	cout<<time<<endl;
	for(int i=0;i<n;i++)
		cout<<a[i]<<" ";
	return 0;
}

File: 算法设计与分析作业/第2章 递归与分治/3.19 快速排序.py
def partition(a,l,r):
    i=l-1
    for j in range(l,r):
        if a[j]<a[r]:
            i+=1
            a[i],a[j]=a[j],a[i]
    a[i+1],a[r]=a[r],a[i+1]
    return i+1

def quicksort(a,l,r):
    global time
    time+=1
    if l<r:
        m=partition(a,l,r)
        quicksort(a,l,m-1)
        quicksort(a,m+1,r)

a=[int(i) for i in input().split(" ")]
n=len(a)
time=0
quicksort(a,0,n-1)
print(time)
for i in range(n):
    print(a[i],end=" ")